### PR TITLE
Refactor destination service and add tests in preparation to add information about labels

### PR DESCRIPTION
**Pull request by pcalcado**
I am making this its own PR so that the follow-up can contain just the new feature without much noise. **The vast majority of LOC here are tests.**

So that we can add the label information for #429, I've made the way we listen to changes on name resolution to be a bit more flexible and added a lot of tests (and eventual refactorings towards testability) to them. I've tested this using the `destination-client` script and it seems to behave the same for the main use cases (IP, external name, kubernetes service).

### CODE REVIEW

File: controller/destination/server_test.go [hunk 1]
```diff
@@ -1,124 +1,232 @@
 package destination
 
 import (
-	"fmt"
+	"context"
+	"reflect"
 	"testing"
 
-	"github.com/stretchr/testify/assert"
+	"github.com/pkg/errors"
```
Comment (klingerf) [inline controller/destination/server_test.go]: It looks like you're importing a 3rd-party lib when you could be using the "errors" package from stdlib. In fact, if you fix this import, then you don't need any of the `Gopkg.*` and deps changes on this branch.

File: controller/destination/k8s_resolver_test.go [hunk 2]
```diff
@@ -0,0 +1,323 @@
...
+		}()
+		cancelFn()
+		<-done
+
+		if mockDnsWatcher.ListenerSubscribed != listener || mockDnsWatcher.ListenerUnsubscribed != listener {
+			t.Fatalf("Expected listener [%v] to have been subscribed then unsubscribed to endpoint watcher, got: %+v", listener, mockEndpointsWatcher)
+		}
+	})
+
+	//TODO: Resolve name using DNS similar to Kubernetes' ClusterFirst
```
[the remark this answers is not in the document]
Comment (pcalcado) [reply controller/destination/k8s_resolver_test.go]: I've tried to implement a test for each comment in the original file: https://github.com/runconduit/conduit/blob/7dc21f9/controller/destination/server.go#L137

I wasn't sure about what to do with this one. Do you think a `Skip` test would work well in documenting this as a missing feature instead?

File: controller/destination/server_test.go (hunk 1 above)
Comment (pcalcado) [reply controller/destination/server_test.go]: That was an intellij import mistake 🤡

**Comment by pcalcado** [comment]
@klingerf I think I got the last few bits!

**Pull request merged**